src/topology/homotopy/H_spaces.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2022 Filippo A. E. Nuccio Mortarino Majno di Capriglio. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Filippo A. E. Nuccio, Junyan Xu
+-/
+import topology.compact_open
+import topology.homotopy.path
+
+/-!
+# H-spaces
+
+This file defines H-spaces mainly following the approach proposed by Serre in his paper
+*Homologie singulière des espaces fibrés*. The main results are the H-space `instance` on every
+topological group, and the H-space structure on the loop space (based at `x : X` of any topological
+space `X`, for which we introduce the notation `Ω_[x]`.
+
+## References
+
+* [J.-P. Serre, *Homologie singulière des espaces fibrés. Applications*,
+  Ann. of Math (2) 1951, 54, 425–505][serre1951]
+-/
+
+universes u v
+
+noncomputable theory
+
+open_locale unit_interval
+
+open path continuous_map set.Icc
+
+/--
+An topological space `X` is an H-space if it behaves like a (potentially non-associative)
+topological group, but where the axioms for a group only hold up to homotopy.
+-/
+class H_space (X : Type u) [topological_space X] :=
+(Hmul : C(X × X, X))
+(e : X)
+(Hmul_e_e : Hmul (e, e) = e)
+(left_Hmul_e : (Hmul.comp $ (const X e).prod_mk $ continuous_map.id X).homotopy_rel
+  (continuous_map.id X) {e})
+(right_Hmul_e : (Hmul.comp $ (continuous_map.id X).prod_mk $ const X e).homotopy_rel
+  (continuous_map.id X) {e})
```

Comment:
These seem to be types. I think they should be props.

Reply:
By "these" do you mean the `(Hmul.comp $ (const X e).prod_mk $ continuous_map.id X).homotopy_rel
  (continuous_map.id X) {e})` (and the similar one on the right)? Then I do not think so, the statement is not simply the mere existence of a potential homotopy, it is the choice of a homotopy from the identity to the multiplication-by-`e`. Or do you mean that you would like them to be existential statements not fixing one particular homotopy?

Reply:
I am not really in favour of rename `e` to `one` due to the "standard"  example of $\mathbb{R}$ being a H-space with $e=z$ for every possible $z\in\mathbb{R}$ and "addition" given by $x\ast y := \frac{x+y}{2}$. A part from this, I can rename them.